sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/UnsafeArrayOutOfBounds.java

```diff
@@ -0,0 +1,554 @@
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "putIntVolatile")
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "putLong",
+      targetMethodDescriptor = "(Ljava/lang/Object;JJ)V")
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "putLong",
+      // Overload with `int offset`, removed in Java 9
+      targetMethodDescriptor = "(Ljava/lang/Object;IJ)V")
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "putLongVolatile")
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "putShort",
+      targetMethodDescriptor = "(Ljava/lang/Object;JS)V")
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "putShort",
+      // Overload with `int offset`, removed in Java 9
+      targetMethodDescriptor = "(Ljava/lang/Object;IS)V")
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "putShortVolatile")
+  @MethodHook(type = HookType.BEFORE, targetClassName = UNSAFE_NAME, targetMethod = "putOrderedInt")
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "putOrderedLong")
+  public static void primitiveSetterHook(
+      MethodHandle method, Object thisObject, Object[] arguments, int hookId) {
+    int accessSize =
+        getBytesCount(method.type().parameterType(2 + 1)); // + 1 for implicit Unsafe instance
+    checkAccess(arguments, accessSize);
+  }
+
+  /** Hook for {@link Unsafe#setMemory(Object, long, long, byte)} */
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "setMemory",
+      // Only handle overload with Object parameter
+      targetMethodDescriptor = "(Ljava/lang/Object;JJB)V")
+  public static void setMemoryHook(
+      MethodHandle method, Object thisObject, Object[] arguments, int hookId) {
+    checkAccess(arguments[0], (long) arguments[1], (long) arguments[2]);
+  }
+
+  /** Hook for {@link Unsafe#copyMemory(Object, long, Object, long, long)} */
+  @MethodHook(
+      type = HookType.BEFORE,
+      targetClassName = UNSAFE_NAME,
+      targetMethod = "copyMemory",
+      // Only handle overload with Object parameters
+      targetMethodDescriptor = "(Ljava/lang/Object;JLjava/lang/Object;JJ)V")
+  public static void copyMemoryHook(
+      MethodHandle method, Object thisObject, Object[] arguments, int hookId) {
+    long size = (long) arguments[4];
+    checkAccess(arguments[0], (long) arguments[1], size);
+    checkAccess(arguments[2], (long) arguments[3], size);
+  }
+
+  private static void report(String message) {
+    Jazzer.reportFindingFromHook(new FuzzerSecurityIssueCritical(message));

```

Comment:
Might be nice to report also for which `Unsafe` method this occurred. Since the hooks are run before the actual `Unsafe` method, it does not appear in the stack trace. For code like `unsafe.putLong(dest, destOffset, unsafe.getLong(src, srcOffset))` that makes it a bit more difficult to tell which of the `Unsafe` calls caused the exception, and you cannot debug inside the `Unsafe` call either since sanitization already happens before.

Any suggestions for how reporting could be improved (in case you consider this an issue) are welcome.